File: fortran/test/test_yamlscript.f90
! Copyright 2023-2026 Ingy dot Net
! This code is licensed under MIT license (See License for details)

! Test the yamlscript Fortran binding.
! load() returns the raw JSON response string from libys, so the
! assertions here check for expected substrings.

program test_yamlscript
  use yamlscript
  implicit none

  type(yamlscript_t) :: ys
  character(len=:), allocatable :: result
  integer :: fails

  fails = 0

  call ys%init()

  ! Load YS code:
  result = ys%load('!ys-0:' // new_line('a') // 'test:: inc(41)')
  call check(index(result, '"test":42') > 0, 'load ys code')

  ! Load plain YAML:
  result = ys%load('foo: bar')
  call check( &
    index(result, '"data"') > 0 .and. index(result, 'bar') > 0, &
    'load plain yaml')

  ! Load invalid input returns an error response:
  result = ys%load(':')
  call check(index(result, '"error"') > 0, 'load error response')

  ! Load multiple times on one instance:
  result = ys%load('!ys-0:' // new_line('a') // 'test:: inc(41)')
  call check(index(result, '"test":42') > 0, 'load multiple times')

  call ys%destroy()

  if (fails > 0) then
    print *, fails, 'test(s) failed'
    stop 1
  end if

contains

  subroutine check(cond, label)
    logical, intent(in) :: cond
    character(len=*), intent(in) :: label

    if (cond) then
      print *, 'ok - ', label
    else
      print *, 'not ok - ', label
      fails = fails + 1
    end if
  end subroutine check

end program test_yamlscript

File: delphi/test/test_yamlscript.pas
{ Copyright 2023-2026 Ingy dot Net
  This code is licensed under MIT license (See License for details) }

{ Test the yamlscript Delphi/Pascal binding. }

program test_yamlscript;

{$mode objfpc}{$H+}

uses
  SysUtils, fpjson, yamlscript;

var
  Fails: Integer = 0;

procedure Test(const Name: string; Cond: Boolean);
begin
  if Cond then
    WriteLn('ok - ', Name)
  else
  begin
    WriteLn('not ok - ', Name);
    Inc(Fails);
  end;
end;

procedure RunTests;
var
  ys: TYAMLScript;
  data: TJSONData;
  obj: TJSONObject;
  threw: Boolean;
begin
  ys := TYAMLScript.Create;
  try
    { Load YS code }
    data := ys.Load('!ys-0:' + LineEnding + 'test:: inc(41)');
    try
      obj := data as TJSONObject;
      Test('load ys code', obj.Integers['test'] = 42);
    finally
      data.Free;
    end;

    { Load plain YAML }
    data := ys.Load('foo: bar');
    try
      obj := data as TJSONObject;
      Test('load plain yaml', obj.Strings['foo'] = 'bar');
    finally
      data.Free;
    end;

    { Load invalid input raises }
    threw := False;
    try
      data := ys.Load(':');
      data.Free;
    except
      on EYAMLScriptException do
        threw := True;
    end;
    Test('load error raises', threw);

    { Load multiple times on one instance }
    data := ys.Load('!ys-0:' + LineEnding + 'test:: inc(41)');
    try
      obj := data as TJSONObject;
      Test('load multiple times', obj.Integers['test'] = 42);
    finally
      data.Free;
    end;
  finally
    ys.Free;
  end;
end;

begin
  RunTests;
  if Fails > 0 then
  begin
    WriteLn(Fails, ' test(s) failed');
    Halt(1);
  end;
end.
